diff --git a/proxy/http/HttpCacheSM.cc b/proxy/http/HttpCacheSM.cc
index 7ad75a5..de6da66 100644
--- a/proxy/http/HttpCacheSM.cc
+++ b/proxy/http/HttpCacheSM.cc
@@ -210,11 +210,24 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
     Debug("http_cache", "[%" PRId64 "] [state_cache_open_write] cache open write failure %d. "
                         "retrying cache open write...",
           master_sm->sm_id, open_write_tries);
-
+#if 0
     open_write(
       &cache_key, lookup_url, read_request_hdr, master_sm->t_state.cache_info.object_read,
       (time_t)((master_sm->t_state.cache_control.pin_in_cache_for < 0) ? 0 : master_sm->t_state.cache_control.pin_in_cache_for),
       retry_write, false);
+#else
+    if (master_sm->t_state.txn_conf->cache_open_write_fail_action != HttpTransact::CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING) {
+      open_write(
+        &cache_key, lookup_url, read_request_hdr, master_sm->t_state.cache_info.object_read,
+        (time_t)((master_sm->t_state.cache_control.pin_in_cache_for < 0) ? 0 : master_sm->t_state.cache_control.pin_in_cache_for),
+        retry_write, false);
+    } else {
+      Debug("http_cache", "[%" PRId64 "] [state_cache_open_write] cache open write failure INTERVAL. Saw CF, not attempting write",
+            master_sm->sm_id);
+      open_write_cb = false;
+      master_sm->handleEvent(CACHE_EVENT_OPEN_WRITE_FAILED, data);
+    }
+#endif
     break;
 
   default:
diff --git a/proxy/http/HttpTransact.cc b/proxy/http/HttpTransact.cc
index 609b59d..14fe639 100644
--- a/proxy/http/HttpTransact.cc
+++ b/proxy/http/HttpTransact.cc
@@ -2404,7 +2404,7 @@ HttpTransact::HandleCacheOpenReadHitFreshness(State *s)
 {
   CacheHTTPInfo *&obj = s->cache_info.object_read;
 
-  ink_release_assert((s->request_sent_time == UNDEFINED_TIME) && (s->response_received_time == UNDEFINED_TIME));
+  //ink_release_assert((s->request_sent_time == UNDEFINED_TIME) && (s->response_received_time == UNDEFINED_TIME));
   DebugTxn("http_seq", "[HttpTransact::HandleCacheOpenReadHitFreshness] Hit in cache");
 
   if (delete_all_document_alternates_and_return(s, true)) {
@@ -3053,6 +3053,24 @@ HttpTransact::handle_cache_write_lock(State *s)
 
       TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
       return;
+    case CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING:
+      if (s->state_machine->get_cache_sm().get_open_write_tries() < s->txn_conf->max_cache_open_write_retries) {
+        // Only log the first CF try, then log the max if hit
+        if (s->state_machine->get_cache_sm().get_open_write_tries() == 1) {
+          DebugTxn("http_error", "collapsed forwarding, cache miss, retry read with collapsed forwarding try #1");
+        }
+        s->request_sent_time       = UNDEFINED_TIME;
+        s->response_received_time  = UNDEFINED_TIME;
+        s->cache_info.action       = CACHE_DO_LOOKUP;
+        s->cache_info.write_status = CACHE_WRITE_LOCK_MISS;
+        break;
+      } else {
+        DebugTxn("http_error", "collapsed forwarding, miss, ran out of write retries (%d), doing default",
+                 s->state_machine->get_cache_sm().get_open_write_tries());
+        s->cache_info.write_status = CACHE_WRITE_LOCK_MISS;
+        remove_ims                 = true;
+        break;
+      }
     default:
       s->cache_info.write_status = CACHE_WRITE_LOCK_MISS;
       remove_ims                 = true;
@@ -3102,6 +3120,13 @@ HttpTransact::handle_cache_write_lock(State *s)
     DebugTxn("http_error", "calling hdr_info.server_request.destroy");
     s->hdr_info.server_request.destroy();
     HandleCacheOpenReadHitFreshness(s);
+  } else if ((s->cache_open_write_fail_action == CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING) &&
+             (s->state_machine->get_cache_sm().get_open_write_tries() < s->txn_conf->max_cache_open_write_retries)) {
+    StateMachineAction_t next;
+    next           = SM_ACTION_CACHE_LOOKUP;
+    s->next_action = next;
+    s->hdr_info.server_request.destroy();
+    TRANSACT_RETURN(next, nullptr);
   } else {
     StateMachineAction_t next;
     if (s->stale_icp_lookup == false) {
diff --git a/proxy/http/HttpTransact.h b/proxy/http/HttpTransact.h
index 3f8bd67..c699327 100644
--- a/proxy/http/HttpTransact.h
+++ b/proxy/http/HttpTransact.h
@@ -304,6 +304,7 @@ public:
     CACHE_WL_FAIL_ACTION_STALE_ON_REVALIDATE               = 0x02,
     CACHE_WL_FAIL_ACTION_ERROR_ON_MISS_STALE_ON_REVALIDATE = 0x03,
     CACHE_WL_FAIL_ACTION_ERROR_ON_MISS_OR_REVALIDATE       = 0x04,
+    CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING              = 0x05,
     TOTAL_CACHE_WL_FAIL_ACTION_TYPES
   };
 
